src/main/java/dao/TaskDaoImpl.java

```diff
@@ -0,0 +1,127 @@
+
+/**
+ * Implementation of the TaskDao using a FileListStorage.
+ */
+public class TaskDaoImpl implements TaskDao {
+	private final Logger logger = Logger.getLogger(TaskDaoImpl.class.getName());
+	
+	/** File containing the ArrayList of Task */
+	private final FileListStorage<Task> taskFileListStorage = new FileListStorage<>(Setting.FILE_TASK_PATH);
+	
+	@Override
+	public void addTask(Task task) {
+		synchronized (this) {
+			try {
+				ArrayList<Task> tasks = taskFileListStorage.readArrayListFromFile();
+				tasks.add(task);
+				taskFileListStorage.writeArrayListToFile(tasks);
+			} catch (Exception e) {
+				logger.warning(e.getMessage());
+			}
+		}
+	}
+	
+	@Override
+	public Task deleteTask(int index) {
+		if (isInvalidIndex(index)) {
+			throw new IndexOutOfBoundsException("non valid index for deletion");
+		}
+		
+		synchronized (this) {
+			try {
+				ArrayList<Task> tasks = taskFileListStorage.readArrayListFromFile();
+				Task removedTask = tasks.remove(index);
+				taskFileListStorage.writeArrayListToFile(tasks);
+				
+				return removedTask;
+			} catch (Exception e) {
+				logger.warning(e.getMessage());
+			}
+		}
+		
+		return null;
+	}
+	
+	@Override
+	public void markDone(int index) {
+		if (isInvalidIndex(index)) {
+			throw new IndexOutOfBoundsException("non valid index for marking the task done");
+		}
+		
+		synchronized (this) {
+			try {
+				ArrayList<Task> tasks = taskFileListStorage.readArrayListFromFile();
+				Task task = tasks.get(index);
+				task.markDone();
+				taskFileListStorage.writeArrayListToFile(tasks);
+			} catch (Exception e) {
+				logger.warning(e.getMessage());
+			}
+		}
+	}
+	
+	@Override
+	public Task getTask(int index) {
+		if (isInvalidIndex(index)) {
+			throw new IndexOutOfBoundsException("non valid index for marking the task done");
+		}
+		
+		try {
+			ArrayList<Task> tasks = taskFileListStorage.readArrayListFromFile();
+			return tasks.get(index);
+		} catch (Exception e) {
+			logger.warning(e.getMessage());
+			return null;
+		}
+	}
+	
+	@Override
+	public List<Task> getAll() {
+		try {
+			return taskFileListStorage.readArrayListFromFile();
+		} catch (Exception e) {
+			logger.warning(e.getMessage());
+			return List.of();
+		}
+	}
+	
+	@Override
+	public List<Task> getByKeyword(String keyword) {
+		try {
+			ArrayList<Task> tasks = taskFileListStorage.readArrayListFromFile();
+			return tasks.stream()
+					.filter(x -> Arrays.asList(x.getDesc().split(" ")).contains(keyword))
+					.collect(Collectors.toList());
+		} catch (Exception e) {
+			logger.warning(e.getMessage());
+			return List.of();
+		}
+	}
+	
+	@Override
+	public int getSize() {
+		try {
+			ArrayList<Task> tasks = taskFileListStorage.readArrayListFromFile();
+			return tasks.size();
+		} catch (Exception e) {
+			logger.warning(e.getMessage());
+			return -1;
+		}
+	}
+	
+	private boolean isInvalidIndex(int index) {
+		int size = getSize();
+		return size != -1 && (index >= getSize() || index < 0);
```

Comment:
Maybe this expression can be simplified for easier understanding

Reply:
I agree, This expression might be confusing to people trying to understand your code.
Maybe you can replace it with:
```
private boolean isInvalidIndex(int index) {
    int size = getSize();
    if (size == -1) {
        return false;
    } else if (index >= size || index < 0) {
        return true;
    } else {
        return false;
    }
}
```
You should also consider moving the error checking for size out of this function.

Reply:
Sure, but I dont think using `if-else` is a good idea since I dont see any benefit from it, ill use more variables that represent better naming instead